pkg/multiboot/ranges.go

```diff
@@ -0,0 +1,261 @@
+
+	"github.com/u-root/u-root/pkg/kexec"
+)
+
+var ErrNotEnoughSpace = errors.New("Not enough space")
+
+var sizeofMmap = uint32(binary.Size(Mmap{}))
+
+type Segments []kexec.Segment
+
+type rangeType string
+
+const (
+	rangeRAM      rangeType = "System RAM"
+	rangeNVACPI             = "ACPI Non-volatile Storage"
+	rangeACPI               = "ACPI Tables"
+	rangeReserved           = "Reserved"
+)
+
+var rangeTypes = map[rangeType]uint32{
+	rangeRAM:      1,
+	rangeNVACPI:   3,
+	rangeACPI:     3,
+	rangeReserved: 4,
+}
+
+type memoryRange struct {
+	start, end uintptr
+	typ        rangeType
+}
+
+type memoryRanges []memoryRange
+
+type Mmap struct {
+	Size uint32
+
+	BaseAddr uint64
+	Length   uint64
+	Type     uint32
+}
+
+type mmaps []Mmap
+
+func (m mmaps) Marshal() ([]byte, error) {
+	buf := bytes.Buffer{}
+	err := binary.Write(&buf, binary.LittleEndian, m)
+	return buf.Bytes(), err
+}
+
+func (m memoryRanges) sub(other memoryRanges) memoryRanges {
+	var ret memoryRanges
+	var i int
+	for _, a := range m {
+		start, end := a.start, a.end
+		for ; i < len(other); i++ {
+			b := other[i]
+			if end < b.start {
+				break
+			}
+			if start < b.start {
+				ret = append(ret, memoryRange{
+					start: start,
+					end:   b.start - 1,
+					typ:   a.typ,
+				})
+			}
+			start = b.end + 1
+		}
+		if start < end {
+			ret = append(ret, memoryRange{
+				start: start,
+				end:   end,
+				typ:   a.typ,
+			})
+		}
+	}
+	return ret
+}
+
+func (m memoryRanges) mmap() mmaps {
+	var ret mmaps
+	for _, r := range m {
+		l := r.end - r.start + 1
+		typ, ok := rangeTypes[r.typ]
+		if !ok {
+			typ = rangeTypes[rangeReserved]
+		}
+		v := Mmap{
+			// Size is really used for skipping to the next pair.
+			Size:     sizeofMmap - 4,
+			BaseAddr: uint64(r.start),
+			Length:   uint64(l),
+			Type:     typ,
+		}
+		ret = append(ret, v)
+	}
+	return ret
+}
+
+func (m memoryRanges) boundaries() (lower, upper uint32) {
+	const M1 = 1048576
+	const K640 = 640 * 1024
+	for _, r := range m {
+		if r.typ != rangeRAM {
+			continue
+		}
+		// Lower memory starts at address 0, and upper memory starts at address 1 megabyte.
+		// The maximum possible value for lower memory is 640 kilobytes.
+		// The value returned for upper memory is maximally the address of the first upper memory hole minus 1 megabyte.
+		// It is not guaranteed to be this value.
+		if r.start <= K640 && uint32(r.end) > lower {
+			lower = uint32(r.end)
+		}
+		if r.start <= M1 && uint32(r.end) > upper+M1 {
+			upper = uint32(r.end - M1)
+		}
+	}
+	return
+}
+
+func (m memoryRanges) filter(typ rangeType) memoryRanges {
+	var ret memoryRanges
+	for _, r := range m {
+		if r.typ == typ {
+			ret = append(ret, r)
+		}
+	}
+	return ret
+}
+
+func (m memoryRanges) findSpace(size uint) (start uintptr, err error) {
+	for _, r := range m {
+		if r.end-r.start >= uintptr(size) {
+			return r.start, nil
+		}
+	}
+	return 0, ErrNotEnoughSpace
+}
+
+func parseMemoryRanges(root string) (memoryRanges, error) {
```

Comment:
This is very Linux-specific, so should be in a file ending with `_linux.go`.

Reply:
done